client_it_test.go

```diff
@@ -92,6 +94,164 @@ func TestClientLifecycleEvents(t *testing.T) {
 	})
 }
 
+func TestClient_AddLifecycleListener(t *testing.T) {
+	receivedStates := []hz.LifecycleState{}
+	receivedStatesMu := &sync.RWMutex{}
+	it.Tester(t, func(t *testing.T, client *hz.Client) {
+		subscriptionID, err := client.AddLifecycleListener(func(event hz.LifecycleStateChanged) {
+			receivedStatesMu.Lock()
+			defer receivedStatesMu.Unlock()
+			// client is already connected
+			switch event.State {
+			case hz.LifecycleStateShuttingDown:
+				t.Logf("Received shutting down state")
+			case hz.LifecycleStateShutDown:
+				t.Logf("Received shut down state")
+			default:
+				t.Log("Received unknown state:", event.State)
+			}
+			receivedStates = append(receivedStates, event.State)
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		assert.NotEqual(t, types.UUID{}, subscriptionID, "subscription UUID should not be empty")
+		if err = client.Shutdown(context.Background()); err != nil {
+			t.Fatal(err)
+		}
+		defer func() {
+			receivedStatesMu.Lock()
+			receivedStates = []hz.LifecycleState{}
+			receivedStatesMu.Unlock()
+		}()
+		targetStates := []hz.LifecycleState{
+			hz.LifecycleStateShuttingDown,
+			hz.LifecycleStateShutDown,
+		}
+		it.Eventually(t, func() bool {
+			receivedStatesMu.RLock()
+			defer receivedStatesMu.RUnlock()
+			return reflect.DeepEqual(targetStates, receivedStates)
+		})
+	})
+}
+
+func TestClient_RemoveLifecycleListener(t *testing.T) {
+	var lifecycleEventReceived int32 = 1
+	it.Tester(t, func(t *testing.T, client *hz.Client) {
+		subscriptionID, err := client.AddLifecycleListener(func(event hz.LifecycleStateChanged) {
+			// shutting down event is published before the actual shutdown is occurred
+			// client must not be shutdown before published that event
+			if event.State == hz.LifecycleStateShuttingDown {
+				atomic.SwapInt32(&lifecycleEventReceived, 0)
+			}
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		assert.NotEqual(t, types.UUID{}, subscriptionID, "subscription UUID should not be empty")
+		if err = client.RemoveLifecycleListener(subscriptionID); err != nil {
+			t.Fatal(err)
+		}
+		if err = client.Shutdown(context.Background()); err != nil {
+			t.Fatal(err)
+		}
+		it.Eventually(t, func() bool {
+			// ensures that client is
+			return !client.Running() && lifecycleEventReceived == 1
```

Comment:
As I explain above, it should stay as 1. Because before the actual shutting down process started, we are removing the subscription to lifecycle event, therefore any lifecycle event won't be received. I assumed that `RemoveLifecycleListener` is a blocking operation. After the line that we call it, we should not have a subscription anymore.